src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java

```diff
@@ -221,28 +221,58 @@ public void doNothing(String mnemonic, Object inputs) throws EvalException {
       SpawnInfo.spawnInfo;
 
   @Override
-  public void symlink(FileApi output, String path) throws EvalException {
+  public void symlink(FileApi output, Object target) throws EvalException {
     context.checkMutable("actions.symlink");
 
+    Artifact outputArtifact = (Artifact)output;
+    // TODO(yannic): This should be configurable.
+    String progressMessage = "Creating symlink " + outputArtifact.getRootRelativePathString();
+
+    if (target instanceof String) {
+      registerUnresolvedSymlinkAction(outputArtifact, (String)target, progressMessage);
+    } else if (target instanceof Artifact) {
+      registerSymlinkAction(outputArtifact, (Artifact)target, progressMessage);
+    } else {
+      // Should have been verified by Starlark before this function is called
+      throw new IllegalStateException();
+    }
+  }
+
+  private void registerSymlinkAction(
+      Artifact output, Artifact input, String progressMessage) throws EvalException {
+    if (output.isSymlink()) {
+      throw Starlark.errorf("output of symlink action must be created by declare_file()");
```

Comment:
nit: "symlinks pointing to artifacts must be created by declare_file()?"

If you want a dangling symlink, it has to created by `declare_symlink()`, doesn't it?